### PR TITLE
Blogapp circleci

### PR DESCRIPTION
**Pull request by Ali-D-Akbar**
# Blog App Backend:

Description: Added the following features:
1. Continuous Integration through CircleCI. Added some configuration files to run `make validate` on every integration.
2. Added some changes related to Pep8 code convention and readability.

Installation Instructions:
1. make a virtual environment by running `python3 -m venv env`
2. Open terminal and start virtual environment by running `source env/bin/activate`
3. run `make requirements` to make requirements
4. run `make serve` to start the django server

Screenshot from 2019-10-22 13-30-18

### CODE REVIEW

File: .circleci/config.yml [hunk 1]
```diff
@@ -0,0 +1,27 @@
+version: 2
+jobs:
+  build:
+    working_directory: ~/circleci-demo-cirpython-django
```
Comment (zubair-arbi) [inline .circleci/config.yml]: working directory name should match your project's name.

File: .circleci/config.yml [hunk 2]
```diff
@@ -0,0 +1,27 @@
...
+      - image: circleci/python:3.6.4
+    steps:
+      - checkout
+      - restore_cache:
+          key: deps1-{{ .Branch }}-{{ checksum "requirements.txt" }}
+      - run:
+          command: |
+            python3 -m venv venv
+            . venv/bin/activate
+            pip install -r requirements.txt
```
Comment (zubair-arbi) [inline .circleci/config.yml]: use the command `make requirements` from `Makefile`

File: .circleci/config.yml [hunk 3]
```diff
@@ -0,0 +1,27 @@
...
+    working_directory: ~/circleci-demo-cirpython-django
+    docker:
+      - image: circleci/python:3.6.4
+    steps:
+      - checkout
+      - restore_cache:
+          key: deps1-{{ .Branch }}-{{ checksum "requirements.txt" }}
+      - run:
+          command: |
+            python3 -m venv venv
```
Comment (zubair-arbi) [inline .circleci/config.yml]: make every line a a run command, e.g.:

```yml
    steps:

      - checkout

      - run: sudo chown -R circleci:circleci /usr/local/bin

      - run: sudo chown -R circleci:circleci /usr/local/lib/python3.6/site-packages

      - run: sudo chown -R circleci:circleci /usr/local/share

      - run: make requirements

      - run: make validate
```

File: blog/models.py [hunk 4]
```diff
@@ -27,8 +31,19 @@ class Meta:
     def votes(self):
         return Blog.objects.aggregate(
             total_votes=(
-                    Count('post_votes', filter=Q(post_votes__vote_type='U') & Q(post_votes__blog=self)) -
-                    Count('post_votes', filter=Q(post_votes__vote_type='D') & Q(post_votes__blog=self))
+                    Count(
+                        'post_votes', filter=(
+                                Q(post_votes__vote_type='U')
```
Comment (zubair-arbi) [inline blog/models.py]:
```python
Q(post_votes__vote_type='U') &
Q(post_votes__blog=self)
```

File: blog/models.py [hunk 5]
```diff
@@ -52,11 +67,19 @@ def downvote(self, user):
...
+        on_delete=models.CASCADE
+    )
+    blog = models.ForeignKey(
+        Blog, related_name='comment', on_delete=models.CASCADE
+    )
     description = models.TextField(max_length=50000)
     created = models.DateTimeField(auto_now_add=True)
-    owner = models.ForeignKey('auth.User', related_name='comment', on_delete=models.CASCADE, null=True)
+    owner = models.ForeignKey(
+        'auth.User', related_name='comment',
```
Comment (zubair-arbi) [inline blog/models.py]: either use one line or add args/kwargs on new lines, e.g.:

```python
owner = models.ForeignKey('auth.User', related_name='comment', on_delete=models.CASCADE, null=True)
```

OR

```python
owner = models.ForeignKey(
    'auth.User', related_name='comment', on_delete=models.CASCADE, null=True
)
```

OR

```python
owner = models.ForeignKey(
    'auth.User', 
    related_name='comment', 
    on_delete=models.CASCADE, 
    null=True
)
```

File: blog/views.py [hunk 6]
```diff
@@ -7,21 +7,16 @@
...
 from blog.models import Blog, Comment, Profile, UserVote
 from blog.permissions import IsOwnerOrReadOnly
-from blog.serializers import (
-    BlogSerializer,
-    CommentSerializer,
-    ProfileSerializer,
-    UserSerializer,
-    VoteSerializer
-)
+from blog.serializers import (BlogSerializer, CommentSerializer,
```
Comment (zubair-arbi) [inline blog/views.py]: use isort for automatically sorting the imports.

Comment (Ali-D-Akbar) [reply blog/views.py]: this one is actually sorted by isort.

File: .circleci/config.yml (hunk 1 above)
Comment (Ali-D-Akbar) [reply .circleci/config.yml]: done.

File: .circleci/config.yml (hunk 2 above)
Comment (Ali-D-Akbar) [reply .circleci/config.yml]: done.

File: .circleci/config.yml (hunk 3 above)
Comment (Ali-D-Akbar) [reply .circleci/config.yml]: Done.

File: blog/models.py (hunk 4 above)
Comment (Ali-D-Akbar) [reply blog/models.py]: Done.

File: .circleci/config.yml [hunk 7]
```diff
@@ -0,0 +1,27 @@
...
+    docker:
+      - image: circleci/python:3.6.4
+    steps:
+      - checkout
+      - restore_cache:
+          key: deps1-{{ .Branch }}-{{ checksum "requirements.txt" }}
+      - run: python3 -m venv venv
+      - run:
+          command: |
+            . venv/bin/activate
```
Comment (zubair-arbi) [inline .circleci/config.yml]: Why we need to activate the env every time?

File: blog/models.py (hunk 5 above)
Comment (Ali-D-Akbar) [reply blog/models.py]: Done.

File: .circleci/config.yml (hunk 7 above)
Comment (Ali-D-Akbar) [reply .circleci/config.yml]: Every run command is independent in CircleCI. It is necessary to run this command if I want to use an installed module.

Comment (Ali-D-Akbar) [reply .circleci/config.yml]: I tried running it without activating again but it didn't work.

**Comment by Ali-D-Akbar** [comment]
The changes have been implemented.

**Pull request merged**